File: leetcode/medium/198_rob.py
from typing import List
from json import dumps
import heapq

class Solution:
    def rob(self, nums: List[int]) -> int:
        robbed = [0]*len(nums)
        robbed[0] = nums[0]
        robbed[1] = nums[1]

        res = max(robbed[0], robbed[1])

        for i in range(2, len(nums)):
            robbed[i] = nums[i]
            for j in range(i-2, -1, -1):
                robbed[i] = max(robbed[i], robbed[j] + nums[i])
                res = max(res, robbed[i])

        print('robbed', robbed)

        return res

def test ():
    params = [
        {
            'input': [1,2,3,1],
            'output': 4,
        },
        {
            'input': [2,7,9,3,1],
            'output': 12,
        },
    ]
    solution = Solution()

    for param in params:
        result = solution.rob(param['input'])
        print(
            'SUCCESS' if result == param['output'] else 'ERROR',
            'input', param['input'],
            'output', param['output'],
            'result', result,
            '\n',
        )


if __name__ == '__main__':
    test()

File: leetcode/medium/198_rob.ts
function rob(nums: number[]): number {
    let prev: number = 0;
    let prevPrev: number = 0;
    for (let i = 0; i < nums.length; ++i) {
        const curr: number = Math.max(prevPrev + nums[i], prev);
        prevPrev = prev;
        prev = curr;
    }

    return prev;
};

const test = () => {
    const params = [
        {
            input: {
                nums: [2,1]
            },
            output: 2,
        },
        {
            input: {
                nums: [0]
            },
            output: 0,
        },
        {
            input: {
                nums: [1,2,3,1]
            },
            output: 4,
        },
        {
            input: {
                nums: [2,7,9,3,1]
            },
            output: 12,
        },

    ];

    params.forEach(({input, output}) => {
        const { nums } = input;
        const result = rob(nums);

        console.log(
            JSON.stringify(result) === JSON.stringify(output) ? 'SUCCESS ' : 'ERROR ',
            'input', JSON.stringify(input),
            'output', output,
            'result', result,
        );
    });
};

test();